Add support for installing ESM module nodes

hlovdal opened the pull request:
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Proposed changes

Currently node-red fails when attempting to install a node written as an ESM module, e.g.

```text
19 Sep 18:59:53 - [info] Server now running at http://127.0.0.1:1880/
19 Sep 18:59:53 - [info] Starting flows
19 Sep 18:59:53 - [info] Started flows
19 Sep 19:01:05 - [info] Installing module: @hlovdal/node-red-lowercase-in-typescript, version: 1.0.0
19 Sep 19:01:12 - [info] Installed module: @hlovdal/node-red-lowercase-in-typescript
19 Sep 19:01:12 - [info] Added node types:
19 Sep 19:01:12 - [info]  - @hlovdal/node-red-lowercase-in-typescript:lower-case : Error [ERR_REQUIRE_ESM]: require() of ES Module .../.node-red/node_modules/@hlovdal/node-red-lowercase-in-typescript/src/lower-case.js from .../node-red/packages/node_modules/@node-red/registry/lib/loader.js not supported.
Instead change the require of lower-case.js in .../node-red/packages/node_modules/@node-red/registry/lib/loader.js to a dynamic import() which is available in all CommonJS modules.
```

This pull request fixes that by using [dynamic `import()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) instead of `require`.

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the [forum](https://discourse.nodered.org/t/adding-support-for-installing-esm-module-nodes/81438)/slack team.
- [x] I have run `npm run test` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

hlovdal committed:
Support loading ESM via dynamic import rather than require

hlovdal committed:
Refactor: extract method createLoadPromise

No variable renames, just lines moved.

hlovdal committed:
Add missing semicolon

hlovdal commented:
I am not sure why the EasyCLA status is not updated, it is fine in [this pull request](https://github.com/node-red/node-red/pull/4356).

hlovdal commented:
/easycla

hlovdal commented:
### TL;DR

The test failure that occurred on my CI build is not related to my changes, it is inherited from the existing code and applies to both `dev` and `master`.

### Test results & conclusion

I have now run testing overnight and the results are:

```bash
$ wc -l success-dev failure-dev
 1312 success-dev
   86 failure-dev
 1398 total
$ wc -l success-installing_esm_nodes failure-installing_esm_nodes
 1297 success-installing_esm_nodes
  103 failure-installing_esm_nodes
 1400 total
$ wc -l success-master failure-master
  416 success-master
   37 failure-master
  453 total
$
```

Or in percentages, the test fails in 6.2%, 7.4% and 8.2% of the times on the various branches, and the total failure rate for all is 7.0%.

The value that fails is almost 200 which it is compared agains, over half is 199 (details in the last step).

So conclusion, the timing granularity/accuracy of this failing test is not good enough and it will fail from time to time (around 7% of the time).


### Details

Step one was to limit testing to just the file that triggered the failure by modifying `Gruntfile.js` to have

```text
            all: { src: ["test/nodes/core/function/10-function_spec.js"]},
            core: { src: ["test/nodes/core/function/10-function_spec.js"]},
            nodes: { src: ["test/nodes/core/function/10-function_spec.js"]}
```

Step two: run the tests through a script that records the success or failure from each test:

```bash
#!/bin/sh

touch success-$1 failure-$1
while true
do
        if npm run test
        then
                date >> success-$1
        else
                date >> failure-$1
        fi
        echo ====================================================
        wc -l success-$1 failure-$1
        echo ====================================================
done
```

Step three was to test on all three branches:

```bash
timeout 8h ./run.sh installing_esm_nodes; git checkout dev; timeout 8h ./run.sh dev; git checkout master; ./run.sh master
```

Step four, analyse failure values.

To be able to see what the failure values were I ran the testing inside a [`script`](https://www.man7.org/linux/man-pages/man1/script.1.html) session, capturing text output to a `typescript` file.

```bash
$ awk '/Message received before init comple/{ A[$9]++ } END {total=0; for (i in A) {total += A[i]} for (i in A) {printf "%d %3d %5.1f%%\n", i, A[i], 100*A[i]/total; }}' typescript 
197   1   0.4%
198  88  38.9%
199 121  53.5%
200  16   7.1%
$
```

(the typescript file includes a few more tests than those captured in the `failure-*` files for those of you double checking the numbers)

hlovdal commented:
Same results with node.js v18.17.1:

```text
  1260 node18-success-installing_esm_nodes
    73 node18-failure-installing_esm_nodes
  1244 node18-success-dev
    90 node18-failure-dev
  1241 node18-success-master
    91 node18-failure-master
  3999 total
```

knolleary commented on packages/node_modules/@node-red/registry/lib/loader.js:
```diff
@@ -344,30 +344,19 @@ function loadNodeSet(node) {
-        }
-        return loadPromise;
+        let importPromise = import(node.file);
+        return importPromise
+            .then(function(module_namespace_object) {
+                // CJS will always have default, ESM might have default. (details in note below)
+                let r = module_namespace_object.default ? module_namespace_object.default : module_namespace_object;
+                // Babel related workaround. (references in note below)
+                r = r.__esModule ? r.default : r;
+                return createLoadPromise(node, r);
+            })
+            .catch(function(err) {
+                node.err = err;
+                return Promise.resolve(node);
+            });
```
```suggestion
        return importPromise
            .then(function(moduleNamespaceObject) {
                // CJS will always have default, ESM might have default. (details in note below)
                let r = moduleNamespaceObject.default ? moduleNamespaceObject.default : moduleNamespaceObject;
                // Babel related workaround. (references in note below)
                r = r.__esModule ? r.default : r;
                return createLoadPromise(node, r);
            })
            .catch(function(err) {
                node.err = err;
                return Promise.resolve(node);
            });
```

fbuys reviewed (commented):
I am excited for this to be released, I left a comment if it helps.

hlovdal commented on packages/node_modules/@node-red/registry/lib/loader.js:
```diff
@@ -344,30 +344,19 @@ function loadNodeSet(node) {
-            loadPromise = Promise.resolve(node);
-        }
-        return loadPromise;
+        let importPromise = import(node.file);
```
The `loadNodeSet` function is supposed to return a promise, which it currently does by chaining an import promise with a load promise. I do not see how an async IIFE is useful here.

hlovdal commented:
The original branch before the rebase is saved in my repo as `installing_esm_nodes.2023`.

Julian-Sz reviewed (approved):
Looking forward to this👍

hlovdal commented:
Argh, the stupid update branch button has a default action to merge dev if you miss the drop down indicator on the side by a few pixels. Of course I intended to rebase. I'll fix it manually with a force push.

hlovdal committed:
Add comment about __esModule handling

hlovdal committed:
Remove unused variables

hlovdal committed:
Fix import url for Windows

Absolute file paths cannot be used on windows, and strictly speaking
using urls is the correct approach even if paths might work,
https://github.com/nodejs/node/issues/34765#issuecomment-673755340.

TotallyInformation commented:
> @TotallyInformation Do you want to be the author of [the commit that fixes the import](https://github.com/hlovdal/node-red/commit/a8816594b9e7d738240658f92662c6ebbd3eba4b)? It is basically identical with [your comment](https://github.com/node-red/node-red/pull/4355#issuecomment-3811382767).

No, that's fine. Thanks for the offer though.

hlovdal commented:
All commits pass after the latest rebase.

```shell
git test add -t test "npm ci && npm run test"
git test run -t test dev..installing_esm_nodes
```

knolleary commented:
@hlovdal realise I missed this for today's beta. My apologies - it's now on my radar to make sure we get it done before final 5.0 release.

knolleary commented:
@hlovdal thank you for your patience. This will be in NR5.
